compiler/qsc_eval/src/lib.rs

```diff
@@ -267,20 +294,61 @@ impl Env {
             .find_map(|scope| scope.bindings.get_mut(&id))
     }
 
-    fn push_scope(&mut self) {
-        self.0.push(Scope::default());
+    fn push_scope(&mut self, frame_id: usize) {
+        let scope = Scope {
+            frame_id,
+            ..Default::default()
+        };
+        self.0.push(scope);
     }
 
     fn leave_scope(&mut self) {
         self.0
             .pop()
             .expect("scope should be entered first before leaving");
     }
+
+    #[must_use]
+    pub fn get_variables_in_top_frame(&self) -> Vec<VariableInfo> {
+        if let Some(scope) = self.0.last() {
+            self.get_variables_in_frame(scope.frame_id)
+        } else {
+            vec![]
+        }
+    }
+
+    #[must_use]
+    pub fn get_variables_in_frame(&self, frame_id: usize) -> Vec<VariableInfo> {
+        let candidate_scopes: Vec<_> = self
+            .0
+            .iter()
+            .filter(|scope| scope.frame_id == frame_id)
+            .map(|scope| scope.bindings.iter())
+            .collect();
+
+        let variables_by_scope: Vec<Vec<VariableInfo>> = candidate_scopes
+            .into_iter()
+            .map(|bindings| {
+                bindings
+                    .map(|(id, var)| VariableInfo {
+                        id: *id,
+                        name: var.name.clone(),
+                        type_name: var.value.type_name().to_string(),
```

Comment:
For complex types, how is the type name formatted? We have custom formatters for UDTs and other complex types in `display.rs` in the language service. They are used in Hover today, and the code should probably be shared so that we get consistent formatting across the editor features.